Feat/asso/membership

AlbanSdl opened the pull request.

AlbanSdl committed:
feat(asso): asso detail page

AlbanSdl committed:
feat: add membership list

AlbanSdl committed:
feat(asso): add role reorder ui

AlbanSdl committed:
feat(asso): role names can be ui modified

AlbanSdl committed:
fix(asso): dnd should consume pointer events on links

AlbanSdl committed:
feat(asso): add button to create role

AlbanSdl committed:
feat(asso): add icons

AlbanSdl committed:
feat(asso): add user selector for future memberAdd ui

AlbanSdl committed:
feat(ui): add code of data modal

AlbanSdl committed:
feat: add style and required properties to modalform

AlbanSdl committed:
feat: add labels for options in ModalForm

AlbanSdl committed:
feat(asso): interface complete

AlbanSdl committed:
fix: bug fix and split page to AssoRole component

AlbanSdl committed:
fix: sort role members

TeddyRoncin commented on src/api/assos/fetchAsso.hook.ts:
```diff
@@ -0,0 +1,14 @@
+      setAsso(body);
+    });
+  }, []);
+  return [asso!, setAsso];
```
par défaut il vaut null, donc ya forcément une erreur là nan ?

TeddyRoncin commented on src/app/assos/[assoId]/page.tsx:
```diff
@@ -0,0 +1,333 @@
+        .flat(),
+    );
+    setPermissions(permissions);
+    console.log('Current user permissions:');
```
console.log

TeddyRoncin commented on src/app/assos/[assoId]/page.tsx:
```diff
@@ -0,0 +1,333 @@
+  };
+
+  const handlePopupSubmit = (data: ModalCallbackType<AssoDetailModalType>) => {
+    if ('user' in data && data.roleId && data.endAt && data.permissions) {
```
J'ai pas trop encore lu comment ça marchait, ya pas moyen de mettre un id sur le modal ? Au lieu de check les clés de l'objet

TeddyRoncin commented on src/app/assos/[assoId]/page.tsx:
```diff
@@ -0,0 +1,333 @@
+
+  return (
+    <Page className={styles.page}>
+      <div className={[styles.headerCard, !asso ? styles.glimmer : ''].filter((i) => i).join(' ')}>
```
pourquoi pas `` `${styles.headerCard} ${!asso ? styles.glimmer : ''}` `` ?

AlbanSdl replied:
Jsuis ptet un peu psychorigide mais j'aime pas les espaces en fin de class="" dans l'inspecteur du navigateur

TeddyRoncin replied:
J'aime pas non plus mais je déteste encore plus la manière dont tu l'as fait, c'est illisible 😂
Par contre à la limite tu peux faire une fonction `buildClassName` ou je ne sais quoi dans un fichier utils si tu préfères

AlbanSdl replied:
Je suis d'avis que le nom va être plus petit, un peu du même style que la TFunction (appelée t absolument partout)
Mais oui l'utilitaire arrive (enfin)

TeddyRoncin commented on src/app/assos/[assoId]/page.tsx:
```diff
@@ -0,0 +1,333 @@
+                setCurrentEditingRole={setCurrentEditingRole}
+              />
+            )}
+            disabled={!editMembersMode || !permissions.has('manage_roles')}></VerticalSortDnd>
```
je l'aurais plutôt appelé `enabled`, ça fait une négation de moins à comprendre

AlbanSdl replied:
J'ai voulu m'aligner sur la façon dont HTML fonctionne (t'as le même attribut disabled)

TeddyRoncin commented on src/app/assos/[assoId]/page.tsx:
```diff
@@ -0,0 +1,333 @@
+  const handlePopupSubmit = (data: ModalCallbackType<AssoDetailModalType>) => {
+    if ('user' in data && data.roleId && data.endAt && data.permissions) {
+      createAssoMember(data.roleId, data.endAt, data.permissions, data.user);
+    } else if ('permissions' in data && extraModalData.memberId && extraModalData.roleId && data.roleId && data.endAt) {
```
Je suis pas sûr que tu le fais, faudrait vider `extraModalData` quand la popup est closed ? (ou quand tu en ouvres une nouvelle qui n'utilise pas de `extraModalData`)
D'ailleurs j'y pense, mais tu pourrais utiliser le `extraModalData` pour stocker l'id du modal ouvert

TeddyRoncin commented on src/app/assos/[assoId]/style.module.scss:
```diff
@@ -0,0 +1,128 @@
+
+.headerCard {
+  background: white;
+  padding: 3ch;
```
tu veux pas utiliser des `em` comme tout le monde ? ^^'
Edit : surtout si en dessous tu utilises aussi des `em`

AlbanSdl replied:
En soit je peux utiliser des em partout mais je trouve ça vla moins parlant de se dire que 1.8em ça correspond peu ou prou à la largeur de 3 caractères...
Alors certes 1ch correspond à la largeur du caractère 0 mais jsuis en mode isok c'est pas trop grand (genre M est beaucoup plus large) et c'est pas trop petit non plus (t est beaucoup plus petit)

TeddyRoncin replied:
On se mettrait donc à utiliser des ch ? En vrai ça me va, mais je suis plutôt pour être cohérent dans les unités. Enfin j'aurais techniquement une petite préférence pour les em, parce que je crois que c'est plus utilisé, mais ch ça me va aussi, j'aime bien ton argument

AlbanSdl replied:
J'ai tout converti en em dans [c7a019e](https://github.com/ungdev/etu-utt-front/pull/31/commits/c7a019e39757f09e1f367241293293c234bca102), je peux me caler sur des em quand je fais mes designs

TeddyRoncin commented on src/components/UI/ModalForm.tsx:
```diff
@@ -0,0 +1,215 @@
+}
+
+// These options are already of type []
+type OptionsFieldsRequired = 'stringList';
```
Ya beaucoup de types, c'est dur à digérer, ça manque de documentation

TeddyRoncin commented on src/components/UI/ModalForm.tsx:
```diff
@@ -0,0 +1,215 @@
+    ? Omit<DataModalEntryBase<T>, 'options'> & { options?: DataModalEntryBase<T>['options'][] }
+    : Omit<DataModalEntryBase<T>, 'options'>;
+
+type DataModalKeys = { [key: string]: keyof DataModalType };
```
J'ai du mal à voir à quoi il sert, quand tu l'utilises tu peux soit utiliser `string[]` (quand tu fais `keyof DataModalKeys`) ou `keyof DataModalType` quand tu fais `DataModalKeys[...]`

AlbanSdl replied:
Il sert à faire disparaitre l'option quand y a pas besoin et en soit, tu parles de `string` mais si j'ai mis un type OptionsFieldsPossible ou OptionsFieldsRequired c'est pour qu'on puisse en rajouter au besoin

AlbanSdl replied:
Ok j'ai mal compris la sélection que t'as faite sur github puisque ça affiche le type du dessus, mais du coup ce type ne sert pas en l'état, il sert uniquement à être extend par un type littéral (je l'ai précisé depuis dans un comment)

TeddyRoncin replied:
Hummm, ok, j'avais pas compris que c'était extended, ma bad

TeddyRoncin commented on src/components/UI/ModalForm.tsx:
```diff
@@ -0,0 +1,215 @@
+type DataModalKeys = { [key: string]: keyof DataModalType };
+
+export type DataModalSchema<T extends DataModalKeys> = {
+  [S in keyof T]: DataModalEntry<T[S]>;
```
Idem ici, je comprends pas à quoi sert `DataModalEntry`, puisque `T[S]` vaut toujours `Date | User | string | string[]`, donc de ce que je comprends, `DataModalEntry<T[S]>` vaut nécessairement `Omit<DataModalEntryBase<T>, 'options'>`

AlbanSdl replied:
Ici T est le type littéral (comme quand tu mets un `as const` après un array). Donc il ne contient que les clés que tu lui as données. D'où le fait que S est la clé littérale et T[S] est donc le type littéral qui est attaché à cette clé (et non `Date | User | string | string[]`)

AlbanSdl replied:
_Hésite pas à ouvrir in ide pour voir ce que ça donne_

TeddyRoncin commented on src/components/UI/ModalForm.tsx:
```diff
@@ -0,0 +1,215 @@
+  submitText: ReactNode;
+};
+
+export type ModalCallbackType<Schema extends DataModalKeys> = { [K in keyof Schema]: DataModalType[Schema[K]] };
```
Mais, c'est un doublon de `ModalStates`, je reconnais x)

TeddyRoncin commented on src/components/UI/ModalForm.tsx:
```diff
@@ -0,0 +1,215 @@
+};
+
+type ModalStates<Schema extends DataModalKeys> = {
+  [K in keyof Schema]: DataModalType[Schema[K]];
```
Idem, j'comprends pas ce type en l'état, les valeurs sont toujours les mêmes

TeddyRoncin commented on src/components/UI/ModalForm.tsx:
```diff
@@ -0,0 +1,215 @@
+
+  // Fade in effect
+  useEffect(() => {
+    requestAnimationFrame(() => setIsHidden(false));
```
C'est pas possible que ça soit déjà visible ?

AlbanSdl replied:
Si mais :
de 1 ce serait une très mauvaise utilisation de la ModalForm (puisque du coup tu modifierais la modal form que tu présentes à l'utilisateur. Déjà ça crée des ui shifts (donc niveau ux c'est 0) et puis t'aurais pu tout afficher dans la première modal puisqu'elle supporte complètement le fait d'être scrollable)
de 2 y aurait juste pas d'animation, c'est pas grave

TeddyRoncin commented on src/components/UI/ModalForm.tsx:
```diff
@@ -0,0 +1,215 @@
+        <div className={styles.container}>
+          <div className={styles.content}>
+            {Object.entries(states).map(([key, state]) => {
+              let variant: ReactNode | null;
```
Nanananana, fais un component pour ça, dans le même fichier si tu veux, mais là t'as un map de quasi 100 lignes avec je ne sais combien d'indentations x)

AlbanSdl replied:
Sadique, tu vas me faire faire encore plus de typage... mais isok défi accepté 😋

TeddyRoncin commented on src/components/UI/VerticalSortDnd.tsx:
```diff
@@ -0,0 +1,120 @@
+  const sensors = useSensors(useSensor(ClickPreservingPointerSensor));
+  const [activeId, setActiveId] = useState<string | null>(null);
+
+  const handleDragStart = (veent: DragStartEvent) => {
```
veent -> event

TeddyRoncin commented on src/components/UI/VerticalSortDnd.tsx:
```diff
@@ -0,0 +1,120 @@
+      onDragEnd={handleDragEnd}
+      autoScroll={true}>
+      <DragOverlay modifiers={[restrictToParentElement]}>
+        <Inflater item={items.find((i) => i.id === activeId)!} />
```
faut pas rajouter un `activeId && (...)` autour du truc ? parce que sinon `items.find((i) => i.id === activeId)!` ça risque de crasher

TeddyRoncin commented on src/components/assos/AssoRole.module.scss:
```diff
@@ -0,0 +1,77 @@
+    padding: 5px;
+    position: absolute;
+    left: -31px;
+    top: -0.05em;
```
en px je pense, parce que là c'est littéralement à peine 1px

TeddyRoncin commented on src/components/assos/AssoRole.tsx:
```diff
@@ -0,0 +1,141 @@
+  const [currentEditingRoleValue, setCurrentEditingRoleValue] = useState<string>(role.name);
+  const { t } = useAppTranslation();
+
+  const userSorter = (a: Member, b: Member) => {
```
Ptit com' pour expliquer le principe du sort, sur quoi on sort ?
Pas évident aussi la façon dont tu l'as écrit, peut-être refactor là avec des if, si c'est pas le plus opti c'est pas grave, yaura jamais un nombre incalculable de membres

TeddyRoncin commented on src/components/assos/AssoRole.tsx:
```diff
@@ -0,0 +1,141 @@
+  editing?: boolean;
+  canEdit: boolean;
+  hasPermission: boolean;
+  hasMembersPermission: boolean;
```
Explicite ici la différence entre `hasPermission`, soit en renamant, soit avec une doc

TeddyRoncin commented on src/components/assos/AssoRole.tsx:
```diff
@@ -0,0 +1,141 @@
+  updateAssoMember: (member: Member, fromRoleId: string) => void;
+  setCurrentEditingRole: (id: string | null) => void;
+}>) {
+  const [currentEditingRoleValue, setCurrentEditingRoleValue] = useState<string>(role.name);
```
`currentEditingRoleName` plutôt

TeddyRoncin commented on src/components/assos/AssoRole.tsx:
```diff
@@ -0,0 +1,141 @@
+                  </div>
+                  {!isOld && canEdit && (
+                    <>
+                      <Button onClick={() => updateAssoMember(member, role.id)} disabled={!hasMembersPermission}>
```
Maybe ajouter `hasMemberPermission` dans la condition, ça sert à rien de l'afficher si on peut dans tous les cas rien faire dans cette `div`

AlbanSdl replied:
Perso j'aime bien afficher les boutons quand même, ça évite que la personne cherche comment faire (en cliquant sur les autres boutons par exemple) alors qu'elle a juste pas la permission. Cela dit oui, on pourrait ajouter un tooltip ou quoi

TeddyRoncin replied:
Tooltip je suis chaud, j'ai pas encore regardé les changes

AlbanSdl replied:
<img width="529" height="157" alt="image" src="https://github.com/user-attachments/assets/e34b94ce-6ce0-4e2f-b74c-da1b8160a8d7" />

(j'ai ajusté la couleur depuis le screen pour match la couleur "disabled")

TeddyRoncin commented on src/components/users/UserCard.module.scss:
```diff
@@ -0,0 +1,38 @@
+    gap: 0.2ch;
+    color: $ung-dark-grey;
+
+    :nth-child(2) {
```
Au lieu d'utiliser ça, tu veux pas mettre une `class` au `div` ? Ca me paraît pas propre (déjà dans le sens où ça veut dire que tu peux pas insérer un truc entre les 2 premières `div` sans qu'il y ait un style par défaut et que ça supprime le style de la nouvelle-3è-`div`)

TeddyRoncin commented on src/components/users/UserSelector.tsx:
```diff
@@ -0,0 +1,60 @@
+  const [lastUpdate, setLastUpdate] = useState<number>(0);
+
+  useEffect(() => {
+    if (Date.now() - lastUpdate >= updateInterval) {
```
Je comprends pas trop l'intérêt du `if` ? Peut-être mettre un autre `useEffect` pour le premier, sinon j'ai l'impression que c'est pas vraiment utile.
Si j'ai arrêté d'écrire depuis 2min, je recommence à écrire, useless de lancer une nouvelle request, faut juste mettre le timeout, à savoir entrer dans le else

AlbanSdl replied:
Ouais en vrai ce sera plus simple comme ça, et en retournant un cleanup qui tej le timeout

TeddyRoncin commented on src/components/users/UserSelector.tsx:
```diff
@@ -0,0 +1,60 @@
+      <div className={styles.resultPool}>
+        {!items.length && <div className={styles.noResult}>{t('users:selector.ui.noResult')}</div>}
+        {items.map(
+          (user: User | null) => user && <UserCard key={user.id} user={user} onSelect={(user) => select(user)} />,
```
`onSelect=select`

TeddyRoncin commented on src/components/users/UserSelector.tsx:
```diff
@@ -0,0 +1,60 @@
+    <div className={styles.userSelector}>
+      <Input
+        value={searchString}
+        onChange={(value) => setSearchString(value)}
```
`onChange=setSearchString`

TeddyRoncin reviewed (commented):
That was long

AlbanSdl committed:
fix: add some doc, c function and random changes

AlbanSdl committed:
fix: add modal id and tweak user selector debounce

AlbanSdl committed:
fix: i was able to uncover a debounce issue due to windows' latence

experimenting ch conversion to em, with a (wrong) basis of 1ch = 0.5em for style clarity

TeddyRoncin commented on src/api/assos/fetchAsso.hook.ts:
```diff
@@ -2,13 +2,13 @@ import { useEffect, useState } from 'react';
 import { Asso } from '@/api/assos/asso.interface';
 
-export function useAsso(assoId: string): [Asso, (asso: Asso) => void] {
+export function useAsso(assoId: string): [Asso | null, (asso: Asso) => void] {
```
`undefined` plutôt ? Par défaut, c'est pas qu'il n'y a pas de valeur, c'est qu'elle est pas définie. Enfin, ce que je fais (je crois, mais je me souviens plus vraiment), c'est que mes hooks comme ça me retournent `<type> | null | undefined`, et `null` est set si j'ai pas réussi à faire la request
Ptet pas intuitif de faire la diff entre les 2, peut-être mettre genre `Asso | NotFetched`, avec NotFetched une enum qui vaut genre "NOT_FETCHED", "NOT_FOUND", ...
Mais j'ai l'impression que ça redécale peut-être le problème du `ApiResult.on(...)`

AlbanSdl replied:
Nah pas la peine de mettre des not fetched/not found
Mais cela dit, t'utilises pas `ApiResult.on('error', () => setMachin(null))` (ouais j'ai fait un ctrl+shift+f sur le projet)

TeddyRoncin replied:
Pas faux

TeddyRoncin commented on src/api/pagination.hook.ts:
```diff
@@ -15,16 +15,19 @@ type PaginationHook<T> = {
 
   const api = useAPI();
 
   const updateItems = (query: Record<string, string>) => {
```
Passer page en param ? Et tu l'ajoutes auto dans query ?

AlbanSdl replied:
Cad, tu parles de séparer page ici en une propriété distincte ? En soit elle est pas censée être plus utilisée que ça ici puisque c'est lors de la mise à jour du filtre, ce serait bizarre de commencer en skippant la page 1. Après c'est quand même possible de préciser la page dans la query mais pas forcément ce qu'on veut je trouve

TeddyRoncin replied:
Yes c'est vrai.
Du coup, si `page` n'est jamais passé, autant ne juste pas considérer qu'il puisse être passé (peut-être même l'interdire ? Comme ça on s'assure qu'il n'y a pas de trucs chelous si jamais).
S'il peut-être passé, je trouve ça bizarre de le passer dans la `query`, il a une valeur différente que tout autre élément de `query` dans la fonction.

AlbanSdl replied:
Je dirais que c'est au développeur de ne pas faire n'importe quoi personnellement. Je considère qu'éventuellement dans certains cas on pourrait vouloir commencer à la page `y` (ex. si une url inclut un numéro de page dans une pagination quelconque)

TeddyRoncin replied:
Et donc pourquoi pas ça ?

> S'il peut-être passé, je trouve ça bizarre de le passer dans la query, il a une valeur différente que tout autre élément de query dans la fonction.

Ce que je veux dire par là, c'est pourquoi pas un autre param que query, pour passer la page, puisque c'est pas un élément classique

TeddyRoncin commented on src/api/pagination.hook.ts:
```diff
@@ -15,16 +15,19 @@ type PaginationHook<T> = {
-    else searching.current = true;
+    let resolvePromise: () => void;
+    if (onUpdate.current) {
+      onUpdate.current.then(() => updateItems(query));
```
Ok, alors si j'ai bien compris, l'idée c'est que si ya 2 update d'affilée, la première tu lances la requête, la deuxième tu queue un truc, qui est call quand la première est finie ? Mais du coup si tu en as 3 ? Ça veut dire que tu effectues quand même la deuxième request, même si elle est pas nécessaire sachant que tu vas faire la troisième juste après.

La manière dont je l'aurais fait (mais il m'a fallu un bon 1/4h pour comprendre ton code, donc ptet qu'il y a un truc qu'il me manque) :

```ts
export function usePaginationLoader<T>(path: string): PaginationHook<T> {
    const nextUpdateQuery = useRef<Record<string, string> | undefined>(undefined);
    const searching = useRef<boolean>(false);
    const runUpdateItemsRequest = (query: Record<string, string>) => {
        searching.current = true;
        api.get().on("success").then(() => {
            if (nextUpdateQuery.current) {
                runUpdateItemsRequest(nextUpdateQuery.current);
                nextUpdateQuery.current = false;
            } else {
                searching.current = false;
            }
        });
    };
    const updateItems = (query: Record<string, string>) => {
        if (searching.current) {
            nextUpdateQuery.current = query;
            return;
        } else {
            runUpdateItemsQuery(query);
        }
    };
}
```

J'ai pas testé, mais j'ai l'impression que c'est moins brainfucking

AlbanSdl replied:
Nan mais en vrai ce qu'il faut surtout c'est abort la requête précédente si elle n'a pas abouti pour renvoyer directement la nouvelle. Ca évite à l'utilisateur de perdre du temps pour rien

TeddyRoncin replied:
Ah bah si tu peux l'abort banger, j'étais parti du principe que tu pouvais pas. Aussi à vérifier que c'est pas possible que l'utilisateur se fasse too many request par le serveur par contre, si tu modifies genre 5 fois par secondes ça monte vite (et puis bon le serv va ptet pas aimer les requêtes en base à l'infini)

AlbanSdl replied:
Tsais qu'on a pas de rate limit ? 👀
Et cela dit c'est pas censé être le cas, puisque t'es censé espacer tes requêtes, comme avant. (directement dans le composant qui gère l'input de la recherche par exemple). La modif était juste pour corriger le comportement si ton internet est vraiment très lent (si la réponse à ta requête ne t'arrive pas avant la requête suivante)

(au hasard c'est windows qui m'a permis de trouver ce comportement buggé)

TeddyRoncin commented on src/components/UI/ModalForm.tsx:
```diff
@@ -7,60 +7,107 @@ import UserSelector from '../users/UserSelector';
-// These options are already of type []
+/**
+ * All keys of {@link DataModalType} whose type require an `options` field in {@link DataModalEntry}.
+ * These are types are already arrays.
```
je pense qu'il y a une typo

TeddyRoncin reviewed (changes requested):
Méga propre

AlbanSdl committed:
fix: add tooltips on disabled button (cf. DisableableButton)

AlbanSdl committed:
fix: pagination hook

AlbanSdl committed:
fix build from previous PRs

* does not build yet, still requires the typescript upgrading coming in PR33

AlbanSdl commented:
Petite précision : je dis que j'ai fix le build, c'est partiellement le cas. La raison pour laquelle ça ne build pas encore est que notre version de typescript n'est pas assez forte pour inférer deux types, ça devrait être réglé par la pr 33.

TeddyRoncin commented on src/api/api.ts:
```diff
@@ -219,7 +239,6 @@ async function internalRequestAPI<RequestType, ResponseType>(
   // Add timeout to the request
-  const abortController = new AbortController();
   const timeout = setTimeout(() => {
     abortController.abort();
```
Peut-être passer une raison `abort("timeout')`, et ensuite tu dois pouvoir la récupérer avec `abortController.signal.reason` dans le `finally`. Comme ça, tu peux différencier un `abort` manuel d'un TO

TeddyRoncin commented on src/app/developers/page.tsx:
```diff
@@ -1,10 +1,8 @@
 
 export default function DeveloperPage() {
-  usePageSettings({});
   return (
```
Profites-en pour mettre le component `<Page>` autour du return ^^'

TeddyRoncin commented on src/components/UI/Tooltip.module.scss:
```diff
@@ -31,6 +33,51 @@
+      }
+    }
+
+    &.size-small {
```
`size_small`, idem pour les autres (sauf s'il y a une raison particulière)

AlbanSdl replied:
Bah en soit, c'est absolument pas idiomatique du css les underscores... Tu mets des tirets dans tes noms de classes normalement

TeddyRoncin replied:
Yes, mais nous on utilise les _ pour que ça soit utilisable comme nom de variable. Enfin on peut changer ça si tu préfères, fondamentalement pourquoi pas, mais dans une autre PR

TeddyRoncin commented on src/components/assos/AssoRole.tsx:
```diff
@@ -128,12 +129,18 @@ export function AssoRole({
-                      </Button>
-                      <Button onClick={() => deleteAssoMember(member.id)} disabled={!hasEditMembersPermission}>
+                      </DisableableButton>
+                      <DisableableButton
```
J'y pense : pourquoi tous les boutons sont pas des disableable buttons ?

AlbanSdl committed:
fix: add seperate failure for request abortion

TeddyRoncin commented on src/api/pagination.hook.ts:
```diff
@@ -38,7 +38,7 @@ export function usePaginationLoader<T>(path: string): PaginationHook<T> {
       })
       .on('fallback', () => delete handler.current)
-      .on(ResponseFailureReason.timeout, () => {}); // Hide toast error here as we might have aborted the request
+      .on(ResponseFailureReason.abort, () => {}); // Hide toast error here as we might have aborted the request
```
tu peux enlever la fin du comment, on a forcément abort la req du coup

AlbanSdl committed:
fix: bad abortion detection

TeddyRoncin committed:
Merge remote-tracking branch 'origin/dev' into feat/asso/membership

TeddyRoncin commented on src/api/pagination.hook.ts:
```diff
@@ -15,51 +15,49 @@ type PaginationHook<T> = {
         lastSearch.current = queryData;
-        searching.current = false;
         pageIndex.current = (page && Number(page)) || 1;
+        delete handler.current;
```
maybe ` handler.current = undefined` instead. Same in all this file

TeddyRoncin committed:
Last comments
